core/trino-main/src/main/java/io/trino/execution/SetRoleTask.java

```diff
@@ -58,20 +57,24 @@ public String getName()
         Session session = stateMachine.getSession();
         Optional<String> catalog = statement.getCatalog()
                 .map(Identifier::getValue);
-        if (catalog.isEmpty()) {
-            throw new TrinoException(NOT_SUPPORTED, "Global roles are not supported yet");
-        }
         catalog.ifPresent(catalogName -> getRequiredCatalogHandle(metadata, session, statement, catalogName));
 
         if (statement.getType() == SetRole.Type.ROLE) {
             String role = statement.getRole().map(c -> c.getValue().toLowerCase(ENGLISH)).orElseThrow();
             if (!metadata.roleExists(session, role, catalog)) {
                 throw semanticException(ROLE_NOT_FOUND, statement, "Role '%s' does not exist", role);
             }
-            accessControl.checkCanSetCatalogRole(SecurityContext.of(session), role, catalog.get());
+            if (catalog.isPresent()) {
+                accessControl.checkCanSetCatalogRole(SecurityContext.of(session), role, catalog.get());
+            }
+            else {
+                if (!metadata.listEnabledRoles(session.getIdentity()).contains(role)) {
+                    denySetRole(role);
+                }
+            }
         }
         SelectedRole.Type type = toSelectedRoleType(statement.getType());
-        stateMachine.addSetRole(catalog.get(), new SelectedRole(type, statement.getRole().map(c -> c.getValue().toLowerCase(ENGLISH))));
+        stateMachine.addSetRole(catalog.orElse("system"), new SelectedRole(type, statement.getRole().map(c -> c.getValue().toLowerCase(ENGLISH))));
```

Comment:
One of the core ideas of this design is we don't change the client protocol, because that takes forever to propagate through all of the client implementations.  Instead, we signal system roles using the system catalog, which is owned by Trino.  For existing clients that support roles, like the JDBC driver, they can simply set roles with `system:my_role`.

As for the QueryStateMachine role fields, I think it is clearer to use the existing method.  If we seperate out this one value of the role map, then we have to add cross checks to the methods which is just a pain, and error prone, without much value.